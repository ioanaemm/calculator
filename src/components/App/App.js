import React, { Component } from "react";
import Header from "../Header/Header";
import Body from "../Body/Body";

class App extends Component {
  constructor() {
    super();
    this.state = {
      firstTerm: "0",
      symbol: "",
      secondTerm: "0"
    };

    this.onSymbolClick = this.onSymbolClick.bind(this);
    this.onNumberClick = this.onNumberClick.bind(this);
    this.onEqualClick = this.onEqualClick.bind(this);
    this.displayNumbers = this.displayNumbers.bind(this);
  }

  onSymbolClick(symbol) {
    this.setState({ symbol });
  }

  onNumberClick(number) {
    if (this.state.symbol === "") {
      if (this.state.firstTerm === "0") {
        this.setState({ firstTerm: String(number) });
      } else {
        this.setState({ firstTerm: this.state.firstTerm + String(number) });
      }
    } else {
      if (this.state.secondTerm === "0") {
        this.setState({ secondTerm: String(number) });
      } else {
        this.setState({ secondTerm: this.state.secondTerm + String(number) });
      }
    }
  }

  onEqualClick() {
    if (this.state.symbol === "+") {
      this.setState({
        firstTerm: String(
          parseInt(this.state.firstTerm) + parseInt(this.state.secondTerm)
        ),
        secondTerm: "0",
        symbol: ""
      });
    } else if (this.state.symbol === "-") {
      this.setState({
        firstTerm: String(
          parseInt(this.state.firstTerm) - parseInt(this.state.secondTerm)
        ),
        secondTerm: "0",
        symbol: ""
      });
    }
  }

  displayNumbers() {
    if (this.state.symbol === "") {
      return this.state.firstTerm;
    } else {
      return this.state.secondTerm;
    }
  }

  render() {
    return (
      <div className="App">
        <Header result={this.displayNumbers()} />
        <Body
          onSymbolClick={this.onSymbolClick}
          onNumberClick={this.onNumberClick}
          onEqualClick={this.onEqualClick}
        />
      </div>
    );
  }
}

export default App;
