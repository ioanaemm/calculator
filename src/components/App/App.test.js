import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import Renderer from "react-test-renderer";

jest.mock("../Header/Header", () => "Header");
jest.mock("../Body/Body", () => "Body");

describe("App component", () => {
  it("renders without crashing", () => {
    const component = Renderer.create(<App />);
    expect(component).toBeDefined();
  });

  it("renders correctly", () => {
    const component = Renderer.create(<App />);
    expect(component).toMatchSnapshot();
  });

  it("sets the symbol", () => {
    const component = Renderer.create(<App />);
    const instance = component.getInstance();
    instance.onSymbolClick("foo");
    expect(instance.state.symbol).toEqual("foo");
  });

  it("handles number when the first term is 0 and when no symbol", () => {
    const component = Renderer.create(<App />);
    const instance = component.getInstance();
    instance.state = {
      symbol: "",
      firstTerm: "0"
    };
    instance.onNumberClick("3");
    expect(instance.state.firstTerm).toEqual("3");
  });

  it("handles number when the first term is not 0 and no symbol", () => {
    const component = Renderer.create(<App />);
    const instance = component.getInstance();
    instance.state = {
      symbol: "",
      firstTerm: "69"
    };
    instance.onNumberClick("6");
    expect(instance.state.firstTerm).toEqual("696");
  });

  it("handles number when the second term is 0 and a symbol", () => {
    // setup
    const component = Renderer.create(<App />);
    const instance = component.getInstance();

    // setting the stage
    instance.state = {
      symbol: "+",
      firstTerm: "345",
      secondTerm: "0"
    };

    // mocks
    const mockSetState = jest.fn();
    instance.setState = mockSetState;

    // code to be tested
    instance.onNumberClick("5");

    // assertions
    expect(mockSetState.mock.calls.length).toEqual(1);
    expect(mockSetState).toBeCalledWith({ secondTerm: "5" });
  });

  it("handles number when there is a symbol", () => {
    const component = Renderer.create(<App />);
    const instance = component.getInstance();
    instance.state = {
      symbol: "+",
      secondTerm: "7"
    };
    instance.onNumberClick("9");
    expect(instance.state.secondTerm).toEqual("79");
  });

  it("handles addition", () => {
    //setup
    const component = Renderer.create(<App />);
    const instance = component.getInstance();

    //setting the stage
    instance.state = {
      symbol: "+",
      firstTerm: "23",
      secondTerm: "56"
    };

    //mocks
    const mockSetState = jest.fn();
    instance.setState = mockSetState;

    //code to be tested
    instance.onEqualClick();

    //assertions
    expect(mockSetState.mock.calls.length).toEqual(1);
    expect(mockSetState).toBeCalledWith({
      symbol: "",
      firstTerm: "79",
      secondTerm: "0"
    });
  });

  it("handles substraction", () => {
    //setup
    const component = Renderer.create(<App />);
    const instance = component.getInstance();

    //setting the stage
    instance.state = {
      symbol: "-",
      firstTerm: "30",
      secondTerm: "23"
    };

    //mocks
    const mockSetState = jest.fn();
    instance.setState = mockSetState;

    //code to be tested
    instance.onEqualClick();

    //assertions
    expect(mockSetState.mock.calls.length).toEqual(1);
    expect(mockSetState).toBeCalledWith({
      symbol: "",
      firstTerm: "7",
      secondTerm: "0"
    });
  });
});
