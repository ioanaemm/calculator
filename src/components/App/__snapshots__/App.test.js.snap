// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
<div
  className="App"
>
  <Header
    result="0"
  />
  <Body
    onEqualClick={[Function]}
    onNumberClick={[Function]}
    onSymbolClick={[Function]}
  />
</div>
`;
